runtime/android/core/src/org/xwalk/core/XWalkActivityDelegate.java

```diff
@@ -104,34 +186,39 @@ public void onActivateFailed() {
         mIsInitializing = false;
 
         if (mXWalkUpdater.updateXWalkRuntime()) {
-            // Set the background to screen_background_dark temporarily if the default background
-            // is null in order to avoid the visual artifacts around the alert dialog
-            Window window = mActivity.getWindow();
-            if (window != null && window.getDecorView().getBackground() == null) {
-                Log.d(TAG, "Set the background to screen_background_dark");
-                window.setBackgroundDrawableResource(android.R.drawable.screen_background_dark);
-                mBackgroundDecorated = true;
+            if (mUpdateListener == null) {
+                // Set the background to screen_background_dark temporarily if
+                // the default background is null in order to avoid the visual
+                // artifacts around the alert dialog
+                Window window = mActivity.getWindow();
+                if (window != null &&
+                        window.getDecorView().getBackground() == null) {
+                    Log.d(TAG, "Set the background to screen_background_dark");
+                    window.setBackgroundDrawableResource(
+                            android.R.drawable.screen_background_dark);
+                    mBackgroundDecorated = true;
+                }
             }
         }
     }
 
     @Override
     public void onActivateCompleted() {
-        if (mDialogManager.isShowingDialog()) mDialogManager.dismissDialog();
-
-        if (mBackgroundDecorated) {
-            Log.d(TAG, "Recover the background");
-            mActivity.getWindow().setBackgroundDrawable(null);
-            mBackgroundDecorated = false;
-        }
-
         mIsInitializing = false;
         mIsXWalkReady = true;
-        mCompleteCommand.run();
-    }
+        if (mUpdateListener == null) {
+            if (mDialogManager.isShowingDialog())
+                mDialogManager.dismissDialog();
+
+            if (mBackgroundDecorated) {
+                Log.d(TAG, "Recover the background");
+                mActivity.getWindow().setBackgroundDrawable(null);
+                mBackgroundDecorated = false;
+            }
 
-    @Override
-    public void onXWalkUpdateCancelled() {
-        mCancelCommand.run();
+            mCompleteCommand.run();
+        } else {
+            mUpdateListener.onXWalkUpdateCompleted();
```

Comment:
This sequence is confusing.
`XWalkUpdater.onXWalkUpdateCompleted -> onActivateCompleted -> XWalkActivityDelegate.onXWalkUpdateCompleted`